src/Collectors/JobCollector.php

```diff
@@ -0,0 +1,81 @@
+<?php
+
+namespace AG\ElasticApmLaravel\Collectors;
+
+use AG\ElasticApmLaravel\Agent;
+use AG\ElasticApmLaravel\Collectors\Interfaces\DataCollectorInterface;
+use Illuminate\Foundation\Application;
+use Illuminate\Queue\Events\JobProcessed;
+use Illuminate\Queue\Events\JobProcessing;
+use Illuminate\Support\Arr;
+use Illuminate\Support\Facades\Log;
+use PhilKra\Events\Transaction;
+use Throwable;
+
+/**
+ * Collects info about the job process
+ */
+class JobCollector extends TimelineDataCollector implements DataCollectorInterface
+{
+    protected $app;
+    protected $agent;
+
+    public function __construct(Application $app, Agent $agent, float $request_start_time)
+    {
+        parent::__construct($request_start_time);
+
+        $this->app = $app;
+        $this->agent = $agent;
+        $this->registerEventListeners();
+    }
+
+    public static function getName(): string
+    {
+        return 'job-collector';
+    }
+
+    protected function registerEventListeners(): void
+    {
+        $this->app->events->listen(JobProcessing::class, function (JobProcessing $event) {
+            $this->startTransaction($this->getTransactionName($event));
+        });
+
+        $this->app->events->listen(JobProcessed::class, function (JobProcessed $event) {
+            $transactionName = $this->getTransactionName($event);
+            $this->addMetadata($transactionName, $event->job);
+            $this->stopTransaction($transactionName);
+        });
+    }
+
+    protected function startTransaction(string $transactionName): Transaction
+    {
+        return $this->agent->startTransaction(
```

Comment:
Will this will start a Job transaction inside the Request transaction? I have to admit that I never tried to do that. I'm curious to see how does it look like.

Reply:
I did a test around that and it got a little funky, but yes it's supposed to be possible. For example, if you do `dispatchNow()`on `Illuminate\Bus\Dispatcher` it runs the job within the same process so you'd start 2 transactions in that case.

I spent a little time trying to build up a $transactionStack so we can set the [parentTransaction](https://github.com/philkra/elastic-apm-php-agent/blob/master/docs/examples/parent-transactions.php) (the existing transactionStore doesn't seem conducive to simply finding the last transaction started). It did make it better but I think my top level transaction looked strange. I have those changes stashed so I can revive them if needed.

Reply:
Controller with dispatchNow(), seems to ignore the job events and middleware entirely, everything falls under the controller transaction. I think that's fine and probably right.
<img width="1578" alt="Screen Shot 2020-02-11 at 12 26 25 PM" src="https://user-images.githubusercontent.com/8845299/74266727-67fd9c00-4cca-11ea-961e-e8421e8914e5.png">

With dispatch() - the controller transaction is empty
<img width="1573" alt="Screen Shot 2020-02-11 at 12 29 37 PM" src="https://user-images.githubusercontent.com/8845299/74266847-98453a80-4cca-11ea-9fc2-7aa35bf8871a.png">

...and the job has everything
<img width="1576" alt="Screen Shot 2020-02-11 at 12 30 10 PM" src="https://user-images.githubusercontent.com/8845299/74266912-bad75380-4cca-11ea-86ea-2dc7b67b8a90.png">

**But if we set up parent transactions...**

Then the job still has everything
<img width="1576" alt="Screen Shot 2020-02-11 at 2 21 36 PM" src="https://user-images.githubusercontent.com/8845299/74275719-080ef180-4cda-11ea-81fe-7fa9f7c71a4b.png">

but the parent (controller) has zero length, and you can't see it but it's 'generic' instead of 'HTTP' 
<img width="1581" alt="Screen Shot 2020-02-11 at 2 21 12 PM" src="https://user-images.githubusercontent.com/8845299/74275673-f594b800-4cd9-11ea-805d-07e12086b8ac.png">

My conclusion is that we should do parent transactions but try to diagnose the 0 transaction timespan and incorrect transaction type.

Reply:
Just an update, I think I did figure out where this needs to go. The problem with the empty one is that the `send()` flushes the transactionStore. So we should only call `collectEvents($name)`, and get rid of `sendTransaction($name)`. But after that it fills out the top transaction but double counts every span:
<img width="1544" alt="Screen Shot 2020-02-11 at 4 51 35 PM" src="https://user-images.githubusercontent.com/8845299/74287259-2206ff00-4cef-11ea-8d1d-7bbf5e44dedb.png">

To fix that, I think the collectors need to store the latest open transaction name on the Measure. Then `collectEvents` would only gather the events for the given name. So each span would reference the innermost transaction as its parent.

I'd like to take a crack at that tomorrow if you think this is progressing in a way that you'll be comfortable with. It makes sense to me... 😂

Reply:
@arkaitzgarro I did spend a good deal of time on this. The transaction stack basically works. But there's pretty significant issues with the time tracking. That mostly centers around the fact that the agent and timeline collector just take in the request time and stick to it. Since the span time is relative to the transaction, each transaction should actually get its own timestamp. This issue seems to flow all the way up to PhilKra though.
I tweaked times left and right until I got it quite close, but not all the way there. It does involve getting rid of REQUEST_TIME_FLOAT and LARAVEL_START because they get in the way, and also requires extending the Transaction object, not unlike LazySpan.
I don't have it in a clean state, most of this is from messing around in the vendor directory. But I could get it into a branch so it's documented at least. Depending what you think, it may be better to support jobs in a simpler state but leave parent transactions somewhat unsupported AKA "known issue". I will say they work pretty nice in Kibana - you can click to go see just the child portion, and then from the child there is a "View Full Trace" button to go back to the parent.

Reply:
@cykirsch I feel your pain with timing, that's also the reason I had to create the `LazySpan` class, because PhilKra events take as start time the moment when the instance was created. I also tried to extend the `Transaction` class, but didn't success that much since base properties are private. I even forked the `elastic-apm-php-agent` repo and start rewriting it, but it was too much...

I'd prefer indeed to start simple, without parent transactions at least to have a way to measure jobs, but knowing that it can be better and can be implemented in the future.